std/socket.zig

```diff
@@ -0,0 +1,676 @@
+                system.ENFILE => return SocketError.SystemFdQuotaExceeded,
+                system.ENOBUFS, system.ENOMEM => return SocketError.SystemResources,
+                system.EPROTONOSUPPORT => return SocketError.ProtocolNotSupported,
+                else => |err| return unexpectedError(err),
+            }
+        } else {
+            const rc = system.socket(domain.toInt(c_int), sock_type.toInt(c_int), try protocol.toInt(c_int));
+            if (rc != InvalidSocketFd) return init(rc);
+            // TODO check for TOO_MANY_OPEN_FILES?
+            switch (WSALastError(-1)) {
+                system.ERROR.WSAEACCES => return SocketError.PermissionDenied,
+                system.ERROR.WSAEAFNOSUPPORT => return SocketError.AddressFamilyNotSupported,
+                system.ERROR.WSAEINVAL => return SocketError.ProtocolFamilyNotAvailable,
+                system.ERROR.WSAEMFILE => return SocketError.ProcessFdQuotaExceeded,
+                system.ERROR.WSAENOBUFS => return SocketError.system.emResources,
+                system.ERROR.WSAEPROTONOSUPPORT, system.ERROR.WSAEPROTOTYPE, system.ERROR.WSAESOCKTNOSUPPORT => return SocketError.ProtocolNotSupported,
+                else => |err| return unexpectedError(@intCast(u32, err)),
+            }
+        }
+    }
+
+    pub fn tcp(domain: Domain) SocketError!Socket {
+        return Socket.open(domain, .Stream, .TCP);
+    }
+
+    pub fn udp(domain: Domain) SocketError!Socket {
+        return Socket.open(domain, .DataGram, Proocol.UDP);
+    }
+
+    pub fn unix(socket_type: SocketType) SocketError!Socket {
+        return Socket.open(.Unix, socket_type, .Unix);
+    }
+
+    pub fn bind(self: Socket, addr: *const net.Address) BindError!void {
+        const rc = system.bind(self.fd, &addr.os_addr, @sizeOf(net.OsAddress));
+
+        if (is_posix) {
+            switch (system.getErrno(@intCast(usize, rc))) {
+                0 => return,
+                system.EACCES => return BindError.AccessDenied,
+                system.EADDRINUSE => return BindError.AddressInUse,
+                system.EBADF => unreachable, // always a race condition if this error is returned
+                system.EINVAL => unreachable,
+                system.ENOTSOCK => unreachable,
+                system.EADDRNOTAVAIL => return BindError.AddressNotAvailable,
+                system.EFAULT => unreachable,
+                system.ELOOP => return BindError.SymLinkLoop,
+                system.ENAMETOOLONG => return BindError.NameTooLong,
+                system.ENOENT => return BindError.FileNotFound,
+                system.ENOMEM => return BindError.system.emResources,
+                system.ENOTDIR => return BindError.NotDir,
+                system.EROFS => return BindError.ReadOnlyFilesystem.em,
+                else => |err| return unexpectedError(err),
+            }
+        } else {
+            switch (WSALastError(rc)) {
```

Comment:
Same thing with all the other WSA functions, only check the error code when you get an non-zero return value.

Reply:
`WSALastError` is not `WSAGetLastError` it's a wrapper that checks the return code and returns `WSAGetLastError` if it non-zero.